x/wasm/keeper/msg_server_integration_test.go

```diff
@@ -60,3 +63,92 @@ func TestStoreCode(t *testing.T) {
 	assert.Equal(t, sender.String(), info.Creator)
 	assert.Equal(t, types.DefaultParams().InstantiateDefaultPermission.With(sender), info.InstantiateConfig)
 }
+
+func TestUpdateAdmin(t *testing.T) {
+	wasmApp := app.Setup(false)
+	ctx := wasmApp.BaseApp.NewContext(false, tmproto.Header{Time: time.Now()})
+
+	var (
+		myAddress       sdk.AccAddress = make([]byte, types.ContractAddrLen)
+		_, _, otherAddr                = testdata.KeyTestPubAddr()
+	)
+
+	specs := map[string]struct {
+		addr           string
+		expErr         bool
+		assertEventsFn func(*testing.T, []abci.Event, string, string)
+	}{
+		"admin can update admin": {
+			addr:   myAddress.String(),
+			expErr: false,
+			assertEventsFn: func(t *testing.T, events []abci.Event, contractAddress, newAdmin string) {
+				assert.Equal(t, 2, len(events))
+				assert.Equal(t, "message", events[0].Type)
+				assert.Equal(t, 2, len(events[0].Attributes))
+				assert.Equal(t, "module", string(events[0].Attributes[0].Key))
+				assert.Equal(t, "wasm", string(events[0].Attributes[0].Value))
+				assert.Equal(t, "sender", string(events[0].Attributes[1].Key))
+				assert.Equal(t, myAddress.String(), string(events[0].Attributes[1].Value))
+				assert.Equal(t, "update_contract_admin", events[1].Type)
+				assert.Equal(t, 2, len(events[1].Attributes))
+				assert.Equal(t, "_contract_address", string(events[1].Attributes[0].Key))
+				assert.Equal(t, contractAddress, string(events[1].Attributes[0].Value))
+				assert.Equal(t, "new_admin_address", string(events[1].Attributes[1].Key))
+				assert.Equal(t, newAdmin, string(events[1].Attributes[1].Value))
+			},
+		},
+		"other address cannot update admin": {
+			addr:   otherAddr.String(),
+			expErr: true,
+		},
+	}
+	for name, spec := range specs {
+		t.Run(name, func(t *testing.T) {
+			_, _, newAdmin := testdata.KeyTestPubAddr()
+
+			// setup
+			storeMsg := &types.MsgStoreCode{
+				Sender:                myAddress.String(),
+				WASMByteCode:          wasmContract,
+				InstantiatePermission: &types.AllowEverybody,
+			}
+			rsp, err := wasmApp.MsgServiceRouter().Handler(storeMsg)(ctx, storeMsg)
+			require.NoError(t, err)
+			var storeCodeResult types.MsgStoreCodeResponse
+			require.NoError(t, wasmApp.AppCodec().Unmarshal(rsp.Data, &storeCodeResult))
+			codeID := storeCodeResult.CodeID
+
+			initMsg := &types.MsgInstantiateContract{
+				Sender: myAddress.String(),
+				Admin:  myAddress.String(),
+				CodeID: codeID,
+				Label:  "test",
+				Msg:    []byte(`{}`),
+				Funds:  sdk.Coins{},
+			}
+			rsp, err = wasmApp.MsgServiceRouter().Handler(initMsg)(ctx, initMsg)
+			require.NoError(t, err)
+
+			var instantiateContractResult types.MsgInstantiateContractResponse
+			require.NoError(t, wasmApp.AppCodec().Unmarshal(rsp.Data, &instantiateContractResult))
+			contractAddress := instantiateContractResult.Address
```

Comment:
This part looks the same, can it be moved outside of the loop?

Reply:
L129 is different in each loop. And, before L129 is needed to unify the state before L129 in each loop.

Reply:
I see, but is it better to use `xCtx, _ := ctx.CacheContext()` in this part, like [this](
https://github.com/170210/wasmd/blob/clearadmin/x/wasm/keeper/msg_server_integration_test.go#L216).
Because I think each spec in this test will set a dummy contract for update.

Reply:
But even without moving, I still think that using cache is necessary, as [this issue](https://github.com/Finschia/wasmd/actions/runs/5876298516/job/15934181729) arose earlier in [this pull request](https://github.com/Finschia/wasmd/pull/70). The fix can be found in this [PR](https://github.com/Finschia/wasmd/pull/77).

Reply:
Looks good! I will use it, too.